File: chapter 3/lab1/c++/src/crc_table.cpp
// Author: zeqiang-lai

#include <iostream>
#include <fstream>
#include <string>
using namespace std;

typedef uint16_t crc;
#define WIDTH  (8 * sizeof(crc))
#define TOPBIT (1 << (WIDTH - 1))
#define POLYNOMIAL 0x1021
// 1 0001 0000 0010 0001

crc crc_table[256];

/**
 * computer remainder of the input bytes stream.
 * @return remainder.
 */
crc crc_remainder(uint8_t msg[], int n_bytes)
{
    crc remainder = 0;
    int index;
    for (int byte = 0; byte < n_bytes; ++byte)
    {
        remainder ^= (msg[byte] << (WIDTH - 8));
        index = remainder >> (WIDTH-8);
        remainder = (remainder << 8) ^ crc_table[index];
    }
    return remainder;
}

/**
 * check the validity of the input bit stream with crc code.
 * @return 1 if it is valid, otherwise 0.
 */
int crc_check(uint8_t msg_with_crc[], int n_bytes)
{
    return (crc_remainder(msg_with_crc, n_bytes) == 0);
}

void build_crc_table()
{
    crc remainder;
    for(int dividend = 0; dividend < 256; ++dividend)
    {
        remainder = dividend << (WIDTH - 8);
        for(uint8_t bit = 8; bit > 0; --bit)
        {		
            if (remainder & TOPBIT)
                remainder = (remainder << 1) ^ POLYNOMIAL;
            else
                remainder = (remainder << 1);
        }
        crc_table[dividend] = remainder;
    }
}

// the code below is the same as the code in crc_bit.cpp.

/**
 * convert binary bit string into bytes stream.
 */
int to_bytes_stream(char const bit_string[], int n_char, uint8_t** bit_stream)
{
    int n_bytes = (n_char+7) / 8;
    *bit_stream = new uint8_t[n_bytes];

    // pad zero at the tail
    // uint8_t tmp_byte;
    // for(int byte=0; byte<n_bytes; ++byte)
    // {
    //     tmp_byte = 0;
    //     for(int bit=0; bit<8; ++bit) {
    //         if(bit_string[(byte+1)*8-bit-1] == '1')
    //             tmp_byte |= 1 << bit;
    //     }
    //     (*bit_stream)[byte] = tmp_byte;
    // }

    // pad zero at the front
    uint8_t tmp_byte;
    for(int byte=n_bytes-1; byte>=0; --byte)
    {
        tmp_byte = 0;
        for(int bit=0; bit<8; ++bit) {
            if(bit_string[n_char-1-(bit+(n_bytes-byte-1)*8)] == '1')
                tmp_byte |= 1 << bit;
        }
        (*bit_stream)[byte] = tmp_byte;
    }

    return n_bytes;
}

/**
 * convert bytes stream into binary bit string.
 */
int to_char_stream(uint8_t const bit_stream[], int n_bytes, char** bit_string)
{
    int n_char = n_bytes*8;
    *bit_string = new char[n_char+1];

    int idx = 0;
    for(int byte=0; byte<n_bytes; ++byte){
        for(int bit=7; bit >=0; bit--) {
            // check if the 'bit'th bit in bit_stream[byte] is 1.
            if(bit_stream[byte] & (1 << bit)) {
                (*bit_string)[idx] = '1';
                idx++;
            } else {
                (*bit_string)[idx] = '0';
                idx++;
            }
        }
    }
    (*bit_string)[idx] = '\0';
    return n_char;
}

int read_configuration(string file_path, string& poly, string& send_string, string& received_string)
{
    // open configuration file
    ifstream inifile(file_path);
    if(inifile.fail()) {
        cerr << "Error: configuration file not found." << endl;
        inifile.close();
        return -1;
    }

    // Parsing, only line 4-6 are used.
    int num_line = 0;
    string line;
    while(getline(inifile, line)) {
        num_line ++;
        switch(num_line) {
            case 4:
                poly = line;
                break;
            case 5:
                send_string = line;
                break;
            case 6:
                received_string = line;
                break;
            default:
                break;
        }
    }
    inifile.close();
    return 0;
}

void show_send_string_result(string send_string)
{
    printf("################# Send String #####################\n");
    uint8_t* msg;
    int len_msg = to_bytes_stream(send_string.c_str(), send_string.length(), &msg);

    //-----------------Computer CRC Code---------------//
    crc crc_code;
    crc_code = crc_remainder((uint8_t*)msg, len_msg);

    printf("Message: %s\n", send_string.c_str());
    printf("Message in Hex: ");
    for(int i=0; i<len_msg; ++i)
        printf("%x", msg[i]);
    printf("\n");
    printf("CRC Code: %x\n", crc_code);

    //--------------Pad the crc code------------------//
    // be careful about the bytes sequence.
    uint8_t* msg_with_crc = new uint8_t[len_msg+sizeof(crc)];
    for(int i=0; i<len_msg; ++i)
        msg_with_crc[i] = msg[i];
    for(int i=0; i<sizeof(crc); ++i)
        msg_with_crc[i+len_msg] = (crc_code >> ((sizeof(crc)-i-1)*8)) % 0x100;
    
    printf("Message with CRC Code in Hex: ");
    for(int i=0; i<len_msg+sizeof(crc); ++i)
        printf("%x", msg_with_crc[i]);
    printf("\n");

    char* bit_string;
    to_char_stream(msg_with_crc, len_msg+sizeof(crc), &bit_string);
    printf("Message with CRC Code in Binary: ");
    printf("%s", bit_string);
    printf("\n");
}


void show_receive_string_result(string receive_string)
{
    printf("################# Receive String #####################\n");
    printf("Message: %s\n", receive_string.c_str());
    if(receive_string.length() <= WIDTH) {
        printf("Invalid receive string: too short, at least 33 bit(32 bit crc code).\n");
        return;
    }

    uint8_t* msg;
    int len_msg = to_bytes_stream(receive_string.c_str(), receive_string.length(), &msg);
    printf("Message in Hex: ");
    for(int i=0; i<len_msg; ++i)
        printf("%x", msg[i]);
    printf("\n");

    crc crc_code = crc_remainder((uint8_t*)msg, len_msg);
    if(crc_code == 0)
        printf("Receive string is valid!\n");
    else
        printf("Invalid Receive string. CRC code=%d\n", crc_code);
}

int main(int argc, char* argv[])
{
    build_crc_table();
    //---------------Read Configuration-------------------//
    string ini_path = "crc.ini";
    if(argc == 2)
        ini_path = string(argv[1]);

    string send_string;
    string received_string;
    string poly;

    if(read_configuration(ini_path, poly, send_string, received_string) != 0)
        return -1;
    
    //--------------show result------------------------//
    show_send_string_result(send_string);
    show_receive_string_result(received_string);

    return 0;
}

File: chapter 3/lab1/java/src/crc_table.java
import java.util.ArrayList;
import java.lang.String;
import java.lang.*;
import java.io.*;
public class crc_bit {
    static final int WIDTH =16;
    static final int TOPBIT =1<<(WIDTH-1);
    static final int POLYNOMIAL =0x1021;
    static String bitstream[];
    static int crc_table[]=new int[256];
    public static int getUint16(int i){
        return i & 0x0000ffff;
    }

    public static void build_crc_table()
    {
        int remainder;
        for(int dividend=0;dividend<256;++dividend)
        {
            remainder=dividend<<(WIDTH-8);
            for(int bit=8;bit>0;--bit)
            {
                if((remainder&TOPBIT)!=0)
                {
                    remainder=(remainder<<1)^POLYNOMIAL;
                }
                else
                {
                    remainder=(remainder<<1);
                }
            }
            crc_table[dividend]=getUint16(remainder);
        }
    }

    public static int crc_remainder(int msg[],int n_bytes)
    {
        int  remainder=0;
        int index;
        for(int bytes=0;bytes<n_bytes;++bytes)
        {
            remainder^=((int)msg[bytes]<<(WIDTH-8));
            index=remainder>>(WIDTH-8);
            remainder=getUint16((remainder<<8)^crc_table[index]);
//            for(int bit=8;bit>0;--bit)
//            {
//                if((remainder&TOPBIT)!=0)
//                { remainder=getUint16(remainder<<1);
//                    remainder=getUint16(remainder^POLYNOMIAL);}
//                else
//                    remainder=(remainder<<1);
//            }
        }
        return remainder;
    }



//    public static boolean crc_check(char msg_with_crc[],int n_bytes)
//    {
//        return(crc_remainder(msg_with_crc,n_bytes)==0);
//    }

   /*public static int to_bytes_stream(ArrayList<Character> bitstring)
    {
        int length=bitstring.size();
        if(length%8!=0)
        {
            int local=(((int)(length/8)+1)*8-length);
            for(int i=local+length;i>length;i--)
            {
                bitstring.set(i,bitstring.get(i-local));
            }
            for(int j=0;j<local;j++)
            {
                bitstring.set(j,'0');
            }
        }
        String st=String.valueOf(bitstring);
        for(int k=0;k<length/8;k++)
        {
           bitstream[k]=Integer.toString(Integer.parseInt(st,2),16);
        }
        return 0;
    }*/

   public static StringBuffer to_bytes_stream(StringBuffer string)
   {
       int length=string.length();
       if(length%8!=0)
       {
           int local=((int)(length/8)+1)*8-length;
           for(int i=0;i<local;i++)
           {
               StringBuffer a=new StringBuffer("0");
               string = a.append(string);
           }
       }
       StringBuffer b=new StringBuffer();
        for(int j=0;j<string.length()/8;j++)
        {
//            String a=Integer.toHexString(Integer.parseInt(string.substring(8*j,8*j+8)));
//            if((Integer.toHexString(Integer.parseInt(string.substring(8*j,8*j+8),2))).length()<2)
//            {
//                StringBuffer a1=new StringBuffer(a);
//                StringBuffer a2=new StringBuffer("0");
//                a1=a2.append(a1);
//                a=a1.toString();
//            }
//            System.out.println(a);
            b.append((Integer.toHexString(Integer.parseInt(string.substring(8*j,8*j+8),2))));
        }
        return b;
   }

    public static StringBuffer read_configuration(String file_path)throws Exception//从ini文件中读取多项式
    {
        StringBuffer sb=new StringBuffer();
        try (FileReader reader = new FileReader(file_path); BufferedReader br = new BufferedReader(reader))
        {
            String line;
            while ((line = br.readLine()) != null)
            {
                sb.append(line);
                sb.append("\n");
            }

        }
        catch (IOException e) {
            System.out.println("Error: configuration file not found.");

        }
        return sb;
    }

    public static void show_send_string_result(StringBuffer send_string)
    {
        System.out.println("################# Send String #####################");
        System.out.print("Message:");
        System.out.println(send_string);
        StringBuffer sendstring2=to_bytes_stream(send_string);
        System.out.print("Message in Hex: ");
        StringBuffer sendstring=new StringBuffer();
        for(int i=0;i<sendstring2.length();i++)
        {
            System.out.print(sendstring2.substring(i,i+1));
        }
        System.out.println("");
        System.out.print("CRC Code: ");
        StringBuffer ans=new StringBuffer(sendstring2.length());
        int ans2[]=new int[sendstring2.length()/2];
        for(int k=0;k<sendstring2.length()/2;k++)
        {
            ans.append(Integer.toBinaryString(Integer.parseInt(sendstring2.substring(2*k,2*k+2),16)));
            ans2[k]=Integer.parseInt(sendstring2.substring(2*k,2*k+2),16);
        }
        int crc_code=crc_remainder(ans2,sendstring2.length()/2);
        System.out.println(Integer.toHexString(crc_code));
        StringBuffer msg_with_code=sendstring2.append(Integer.toHexString(crc_code));
        System.out.print("Message With Code in Hex:");
        for(int i=0;i<msg_with_code.length();i++)
        {
            System.out.print(msg_with_code.substring(i,i+1));
        }
        System.out.println("");
    }

    public static void show_receive_string_result(StringBuffer receive_string)
    {
        System.out.println("################# Receive String #####################");
        System.out.print("Message_Received:");
        System.out.println(receive_string);
        System.out.print("Message in Hex: ");
        StringBuffer receivestring2=to_bytes_stream(receive_string);
        StringBuffer receivestring=new StringBuffer();
        for(int i=0;i<receivestring2.length();i++)
        {
            System.out.print(receivestring2.substring(i,i+1));
        }
        StringBuffer ans=new StringBuffer(receivestring2.length());
        int ans2[]=new int[receivestring2.length()/2];
        for(int k=0;k<receivestring2.length()/2;k++)
        {
            ans.append(Integer.toBinaryString(Integer.parseInt(receivestring2.substring(2*k,2*k+2),16)));
            ans2[k]=Integer.parseInt(receivestring2.substring(2*k,2*k+2),16);
        }
        int crc_code=crc_remainder(ans2,receivestring2.length()/2);
        System.out.println(" ");
        if (crc_code==0)
        {
            System.out.println("Receive string is valid!");
        }
        else
        {
            System.out.print("Invalid Receive string. CRC code=");
            System.out.println(Integer.toHexString(crc_code));
        }
    }

    public static void main(String args[])throws Exception
    {
//        String a="hello";
//        char a2[]=a.toCharArray();
//        System.out.println(crc_remainder(a2,5));
        build_crc_table();
        String ini_path="crc.ini";
        String h[]=new String[100];//存储划分后的子串
        if(args.length==1)
        {
            ini_path=args[0];
        }
        int i=0;
        StringBuffer ans=read_configuration(ini_path);
        String[]lines=ans.toString().split("\\n");
        for(String line:lines)//按换行符对读出的字符串进行划分
        {
            h[i]=line;
            i++;
        }
        StringBuffer poly=new StringBuffer(h[3]);
        StringBuffer send_string =new StringBuffer(h[4]);
        StringBuffer recevied_string =new StringBuffer(h[5]);
        recevied_string=new StringBuffer(h[5]);
        send_string =new StringBuffer(h[4]);
        show_send_string_result(send_string);
        show_receive_string_result(recevied_string);
    }
}